File: mangledotdev-v0.1.0_Ruby/mangledotdev.rb
require 'json'
require 'open3'
require 'securerandom'
require 'stringio'

# InputManager - Manages sending requests to other processes and handling responses.
#
# This is an instance-based class - create one instance per request.
#
# Attributes:
#   response (Hash): Complete response with status, data, errors, warnings
#
# Methods:
#   request(): Send a request to another process
#   get_response(): Get the full response object
#   get_data(): Get the response data (returns nil on error)
class InputManager
  attr_reader :response

  # Generate a unique key for request/response matching.
  # @return [String] Unique key
  def self.gen_key
    SecureRandom.hex(16)
  end

  # Initialize a new InputManager instance.
  def initialize
    @process = nil
    @raw_request = nil
    @request = nil
    @response = nil
    @data = nil
    @key = nil
  end

  private

  # Validate file and build command to execute.
  #
  # @param language [String] Programming language/runtime
  # @param file [String] Path to file to execute
  # @return [Array<String>] Command array for subprocess
  # @raise [ArgumentError] Invalid file extension, file not found, or permission error
  def get_command(language, file)
    lang_upper = language.to_s.upcase

    # On Windows, convert forward slashes to backslashes for file system operations
    if Gem.win_platform?
      file = file.gsub('/', '\\')
    end

    file_ext = File.extname(file).downcase

    # Extension validation - FIRST before file existence check
    extension_map = {
      'PYTHON' => ['.py'],
      'PY' => ['.py'],
      'JAVASCRIPT' => ['.js'],
      'JS' => ['.js'],
      'NODE' => ['.js'],
      'NODEJS' => ['.js'],
      'RUBY' => ['.rb'],
      'RB' => ['.rb'],
      'C' => ['.c', '.out', '.exe', ''],
      'CS' => ['.exe', '.dll', ''],
      'CPP' => ['.cpp', '.cc', '.cxx', '.out', '.exe', ''],
      'C#' => ['.exe', '.dll', ''],
      'C++' => ['.cpp', '.cc', '.cxx', '.out', '.exe', ''],
      'CSHARP' => ['.exe', '.dll', ''],
      'CPLUSPLUS' => ['.cpp', '.cc', '.cxx', '.out', '.exe', ''],
      'EXE' => ['.cpp', '.cc', '.cxx', '.out', '.exe', ''],
      'JAR' => ['.jar'],
      'JAVA' => ['.jar'],
      'RUST' => ['.rs', '.exe', '.out', ''],
      'RS' => ['.rs', '.exe', '.out', ''],
      'GO' => ['.go', '.exe', '.out', ''],
      'GOLANG' => ['.go', '.exe', '.out', '']
    }

    if extension_map.key?(lang_upper)
      valid_extensions = extension_map[lang_upper]
      unless valid_extensions.include?(file_ext)
        expected = valid_extensions.map { |ext| ext.empty? ? '(no extension)' : ext }.join(', ')
        raise ArgumentError, "Invalid file '#{file}' for language '#{language}'. Expected: e.g. 'file#{expected}'"
      end
    end

    # File existence check
    raise ArgumentError, "File not found: #{file}" unless File.exist?(file)
    raise ArgumentError, "Path is not a file: #{file}" unless File.file?(file)

    # Permission checks
    compiled_languages = ['C', 'CS', 'CPP', 'C#', 'C++', 'CSHARP', 'CPLUSPLUS', 'EXE', 'RUST', 'RS', 'GO', 'GOLANG']
    if compiled_languages.include?(lang_upper)
      raise ArgumentError, "File is not executable: #{file}" unless File.executable?(file)
    else
      raise ArgumentError, "File is not readable: #{file}" unless File.readable?(file)
    end

    # Auto-add ./ for compiled executables if not present and not absolute path
    if compiled_languages.include?(lang_upper) && !File.absolute_path?(file) && !file.start_with?('./')
      file = './' + file
    end

    # Build command
    lang_map = {
      'PYTHON' => ['python', file],
      'PY' => ['python', file],
      'JAVASCRIPT' => ['node', file],
      'JS' => ['node', file],
      'NODE' => ['node', file],
      'NODEJS' => ['node', file],
      'RUBY' => ['ruby', file],
      'RB' => ['ruby', file],
      'C' => [file],
      'CS' => file_ext == '.dll' ? ['dotnet', file] : [file],
      'CPP' => [file],
      'C#' => file_ext == '.dll' ? ['dotnet', file] : [file],
      'C++' => [file],
      'CSHARP' => file_ext == '.dll' ? ['dotnet', file] : [file],
      'CPLUSPLUS' => [file],
      'EXE' => [file],
      'JAR' => ['java', '-jar', file],
      'JAVA' => ['java', '-jar', file],
      'RUST' => [file],
      'RS' => [file],
      'GO' => file_ext == '.go' ? ['go', 'run', file] : [file],
      'GOLANG' => file_ext == '.go' ? ['go', 'run', file] : [file]
    }

    raise ArgumentError, "Unsupported language: #{language}" unless lang_map.key?(lang_upper)

    lang_map[lang_upper]
  end

  public

  # Send a request to another process.
  #
  # @param is_unique [Boolean] Expect single output (true) or multiple (false)
  # @param optional_output [Boolean] Output is optional (true) or required (false)
  # @param data [Object] Data to send (any JSON-serializable type)
  # @param language [String] Target language/runtime
  # @param file [String] Path to target file
  #
  # Sets @response (Hash) with keys:
  #     - request_status (Boolean|nil): Success status
  #     - data: Response data (preserves type)
  #     - optionalOutput (Boolean): Echo of parameter
  #     - isUnique (Boolean): Echo of parameter
  #     - warnings (Array<String>): Warning messages
  #     - errors (Array<String>): Error messages
  def request(is_unique: true, optional_output: true, data: nil, language:, file:)
    begin
      @key = InputManager.gen_key
      command = get_command(language, file)

      @raw_request = {
        key: @key,
        optionalOutput: optional_output,
        isUnique: is_unique,
        data: data
      }
      @request = JSON.generate(@raw_request)

      response = {
        request_status: nil,
        data: nil,
        optionalOutput: optional_output,
        isUnique: is_unique,
        warnings: [],
        errors: []
      }

      stdout, stderr, status = Open3.capture3(*command, stdin_data: @request)

      unless status.success?
        response[:request_status] = false
        response[:errors] << "Process exited with code #{status.exitstatus}"
        response[:errors] << "stderr: #{stderr.strip}" unless stderr.strip.empty?
        response[:warnings] << "Warning: these kind of errors result from an error in the targeted script."
        @response = response
        return
      end

      @response_data = []
      stdout.strip.split("\n").each do |line|
        next if line.strip.empty?

        begin
          parsed_data = JSON.parse(line)

          # Validate response has matching key or null key (for init errors)
          # This ensures we only process responses meant for this request
          if parsed_data.is_a?(Hash) && (parsed_data['key'] == @key || parsed_data['key'].nil?)
            @response_data << parsed_data
          end
        rescue JSON::ParserError
          # Ignore lines that aren't valid JSON (e.g., debug prints)
        end
      end

      if @response_data.length != 0
        failure = false
        @response_data.each do |resp|
          failure = true unless resp['request_status']
        end

        response[:request_status] = !failure

        response[:isUnique] = @response_data[0]['isUnique']

        @response_data.each do |resp|
          resp['errors'].each do |err|
            response[:errors] << err
          end
        end

        data_list = []
        @response_data.each do |resp|
          data_list << resp['data']
        end

        if response[:isUnique]
          if data_list.length == 1
            response[:data] = data_list[0]
          else
            response[:request_status] = false
            response[:data] = nil
            response[:errors] << "Error: Expected 1 output (isUnique=True) but received #{data_list.length}."
          end
        else
          response[:data] = data_list
        end
      elsif optional_output
        response[:request_status] = nil
        response[:data] = nil
        response[:warnings] << "Warning: the output setting is set to optional, and the targeted program didn't gave any output."
      else
        response[:request_status] = false
        response[:data] = nil
        response[:errors] << "Error: OutputManager might not be used or not correctly."
      end

      @response = response

    rescue ArgumentError => e
      @response = {
        request_status: false,
        data: nil,
        optionalOutput: optional_output,
        isUnique: is_unique,
        warnings: ["Warning: targeted file not found or can't be executed, consider checking file informations and language dependencies."],
        errors: ["Error: #{e.message}"]
      }
    rescue => e
      @response = {
        request_status: false,
        data: nil,
        optionalOutput: optional_output,
        isUnique: is_unique,
        warnings: [],
        errors: ["Unexpected error: #{e.message}"]
      }
    end
  end

  # Get the full response object.
  #
  # @return [Hash] The complete response with status, data, errors, warnings.
  def get_response
    @response
  end

  # Get the response data if request was successful.
  #
  # @return [Object, nil] The data from the response (any type), or nil if request failed.
  #                       The return type matches the type sent by the target process.
  def get_data
    if @response
      return @response[:data] if @response[:request_status]
    end
    nil
  end

  # Bundle any value for use with request() - for API consistency with other languages.
  # In Ruby, this just returns the value as-is since Ruby handles serialization automatically.
  #
  # @param value [Object] Any value
  # @return [Object] The same value (Ruby handles JSON serialization automatically)
  def self.bundle(value)
    value
  end
end

# OutputManager - Manages receiving requests from other processes and sending responses.
#
# This is a class-based/static manager - all methods are class methods.
# Must call init() before using.
#
# Class Attributes:
#   data: The request data (accessible after init())
#
# Class Methods:
#   init(): Initialize and read request from stdin
#   output(val): Send response back via stdout
class OutputManager
  @@original_stdout = nil
  @@request = nil
  @@data = nil
  @@request_status = nil
  @@optional = nil
  @@unique_state = nil
  @@init_error = nil
  @@errors = []
  @@warnings = []

  class << self
    attr_accessor :data

    # Initialize OutputManager and read request from stdin.
    #
    # Must be called before using output() or accessing data.
    # Suppresses stdout to prevent pollution of JSON protocol.
    def init
      # Save original stdout so we can restore it later
      @@original_stdout = $stdout

      # Redirect stdout to StringIO to suppress all puts/print statements
      $stdout = StringIO.new

      # Read the entire stdin (the JSON request from InputManager)
      @@request = $stdin.read
      @@data = JSON.parse(@@request)
      self.data = @@data['data']
      @@optional = @@data['optionalOutput']

      # Reset state for new request
      @@errors = []
      @@warnings = []
      @@init_error = nil
      @@request_status = nil
      @@unique_state = nil
    end

    # Send response back to the calling process.
    #
    # @param val [Object] Data to send (any JSON-serializable type)
    #
    # Note:
    #   Can be called multiple times if isUnique=false in request.
    #   Will error if called multiple times when isUnique=true.
    def output(val)
      # Check if OutputManager was initialized
      if @@data.nil?
        unless @@init_error
          $stdout = @@original_stdout if @@original_stdout

          @@request_status = false
          @@errors << "Error: OutputManager isn't initialized."
          write_output(nil, @@data)
          @@init_error = true
        end
      else
        # Check if we can output based on isUnique setting
        # unique_state tracks if we've already output once
        if !@@unique_state || !@@data['isUnique']
          @@request_status = true
          write_output(val, @@data)
        else
          # Multiple outputs when isUnique=true is an error
          @@request_status = false
          @@errors << "Error: outputs out of bound (isUnique: #{@@unique_state})."
          write_output(val, @@data)
        end

        # Mark that we've output once
        @@unique_state = @@data['isUnique']

        # Re-suppress stdout after writing response
        $stdout = StringIO.new
      end
    end

    # Bundle any value for use with output() - for API consistency with other languages.
    # In Ruby, this just returns the value as-is since Ruby handles serialization automatically.
    #
    # @param value [Object] Any value
    # @return [Object] The same value (Ruby handles JSON serialization automatically)
    def bundle(value)
      value
    end

    private

    # Internal method to write JSON response to stdout.
    #
    # @param args [Object] Data to send in response
    # @param _data [Hash] Request data for key/metadata
    def write_output(args, _data)
      # Restore original stdout to actually write the response
      $stdout = @@original_stdout if @@original_stdout

      # Build and write JSON response
      response = {
        key: _data ? _data['key'] : nil,
        request_status: @@request_status,
        data: args,
        optionalOutput: @@optional,
        isUnique: _data ? _data['isUnique'] : nil,
        errors: @@errors,
        warnings: @@warnings
      }

      puts JSON.generate(response)
      $stdout.flush
    end
  end
end

File: JavaScript/mangledotdev.js
const { spawnSync } = require('child_process');
const fs = require('fs');
const path = require('path');

/**
 * InputManager - Manages sending requests to other processes and handling responses.
 * 
 * This is an instance-based class - create one instance per request.
 * 
 * @property {Object} response - Complete response with status, data, errors, warnings
 */
class InputManager {
    #process;
    #rawRequest;
    #request;
    #responseData;
    #key;

    /**
     * Generate a unique key for request/response matching.
     * @private
     * @returns {string} Unique key
     */
    static #genKey() {
        return [...Array(32)]
        .map(() => (Math.random() * 256 | 0).toString(16).padStart(2, '0'))
        .join('');
    }

    /**
     * Initialize a new InputManager instance.
     */
    constructor() {
        this.#process = null;
        this.#rawRequest = null;
        this.#request = null;
        this.#responseData = null;
        this.response = null;
        this.#key = null;
    }

    /**
     * Validate file and build command to execute.
     * 
     * @private
     * @param {string} language - Programming language/runtime
     * @param {string} file - Path to file to execute
     * @returns {Array<string>} Command array for subprocess
     * @throws {Error} Invalid file extension, file not found, or permission error
     */
    #getCommand(language, file) {
        const langUpper = String(language).toUpperCase();

        // On Windows, convert forward slashes to backslashes for file system operations
        if (process.platform === 'win32') {
            file = file.replace(/\//g, '\\');
        }

        const fileExt = path.extname(file).toLowerCase();

        // Extension validation - FIRST before file existence check
        const extensionMap = {
            'PYTHON': ['.py'],
            'PY': ['.py'],
            'JAVASCRIPT': ['.js'],
            'JS': ['.js'],
            'NODE': ['.js'],
            'NODEJS': ['.js'],
            'RUBY': ['.rb'],
            'RB': ['.rb'],
            'C': ['.c', '.out', '.exe', ''],
            'CS': ['.exe', '.dll', ''],
            'CPP': ['.cpp', '.cc', '.cxx', '.out', '.exe', ''],
            'C#': ['.exe', '.dll', ''],
            'C++': ['.cpp', '.cc', '.cxx', '.out', '.exe', ''],
            'CSHARP': ['.exe', '.dll', ''],
            'CPLUSPLUS': ['.cpp', '.cc', '.cxx', '.out', '.exe', ''],
            'EXE': ['.cpp', '.cc', '.cxx', '.out', '.exe', ''],
            'JAR': ['.jar'],
            'JAVA': ['.jar'],
            'RUST': ['.rs', '.exe', '.out', ''],
            'RS': ['.rs', '.exe', '.out', ''],
            'GO': ['.go', '.exe', '.out', ''],
            'GOLANG': ['.go', '.exe', '.out', '']
        };

        if (langUpper in extensionMap) {
            const validExtensions = extensionMap[langUpper];
            if (!validExtensions.includes(fileExt)) {
                const expected = validExtensions.map(ext => ext ? ext : '(no extension)').join(', ');
                throw new Error(`Invalid file '${file}' for language '${language}'. Expected: e.g. 'file${expected}'`);
            }
        }

        // File existence check
        if (!fs.existsSync(file)) {
            throw new Error(`File not found: ${file}`);
        }

        const stats = fs.statSync(file);
        if (!stats.isFile()) {
            throw new Error(`Path is not a file: ${file}`);
        }

        // Permission checks
        const compiledLanguages = ['C', 'CS', 'CPP', 'C#', 'C++', 'CSHARP', 'CPLUSPLUS', 'EXE', 'RUST', 'RS', 'GO', 'GOLANG'];
        if (compiledLanguages.includes(langUpper)) {
            try {
                fs.accessSync(file, fs.constants.X_OK);
            } catch (err) {
                throw new Error(`File is not executable: ${file}`);
            }
        } else {
            try {
                fs.accessSync(file, fs.constants.R_OK);
            } catch (err) {
                throw new Error(`File is not readable: ${file}`);
            }
        }

        // Auto-add ./ for compiled executables if not present and not absolute path
        if (compiledLanguages.includes(langUpper) && !path.isAbsolute(file) && !file.startsWith('./')) {
            file = './' + file;
        }

        // Build command
        const langMap = {
            'PYTHON': ['python', file],
            'PY': ['python', file],
            'JAVASCRIPT': ['node', file],
            'JS': ['node', file],
            'NODE': ['node', file],
            'NODEJS': ['node', file],
            'RUBY': ['ruby', file],
            'RB': ['ruby', file],
            'C': [file],
            'CS': fileExt === '.dll' ? ['dotnet', file] : [file],
            'CPP': [file],
            'C#': fileExt === '.dll' ? ['dotnet', file] : [file],
            'C++': [file],
            'CSHARP': fileExt === '.dll' ? ['dotnet', file] : [file],
            'CPLUSPLUS': [file],
            'EXE': [file],
            'JAR': ['java', '-jar', file],
            'JAVA': ['java', '-jar', file],
            'RUST': [file],
            'RS': [file],
            'GO': fileExt === '.go' ? ['go', 'run', file] : [file],
            'GOLANG': fileExt === '.go' ? ['go', 'run', file] : [file]
        };

        if (!(langUpper in langMap)) {
            throw new Error(`Unsupported language: ${language}`);
        }

        return langMap[langUpper];
    }

    /**
     * Send a request to another process.
     * 
     * @param {boolean} [isUnique=true] - Expect single output (true) or multiple (false)
     * @param {boolean} [optionalOutput=true] - Output is optional (true) or required (false)
     * @param {*} [data=null] - Data to send (any JSON-serializable type)
     * @param {string} language - Target language/runtime
     * @param {string} file - Path to target file
     * 
     * Sets this.response with keys:
     *     - request_status (boolean|null): Success status
     *     - data: Response data (preserves type)
     *     - optionalOutput (boolean): Echo of parameter
     *     - isUnique (boolean): Echo of parameter
     *     - warnings (Array<string>): Warning messages
     *     - errors (Array<string>): Error messages
     */
    request(isUnique = true, optionalOutput = true, data = null, language, file) {
        try {
            this.#key = InputManager.#genKey();
            const command = this.#getCommand(language, file);

            this.#rawRequest = {
                key: this.#key,
                optionalOutput: optionalOutput,
                isUnique: isUnique,
                data: data
            };
            this.#request = JSON.stringify(this.#rawRequest);

            const response = {
                request_status: null,
                data: null,
                optionalOutput: optionalOutput,
                isUnique: isUnique,
                warnings: [],
                errors: []
            };

            const [cmd, ...args] = command;
            const result = spawnSync(cmd, args, {
                input: this.#request,
                encoding: 'utf-8'
            });

            const stdout = result.stdout || '';
            const stderr = result.stderr || '';
            const code = result.status;

            if (code !== 0) {
                response.request_status = false;
                response.errors.push(`Process exited with code ${code}`);
                if (stderr.trim()) {
                    response.errors.push(`stderr: ${stderr.trim()}`);
                }
                response.warnings.push("Warning: these kind of errors result from an error in the targeted script.");
                this.response = response;
                return;
            }

            this.#responseData = [];
            const lines = stdout.trim().split('\n');
            
            for (const line of lines) {
                if (!line.trim()) {
                    continue;
                }
                
                try {
                    const parsedData = JSON.parse(line);
                    
                    // Validate response has matching key or null key (for init errors)
                    // This ensures we only process responses meant for this request
                    if (typeof parsedData === 'object' && 
                        parsedData !== null &&
                        (parsedData.key === this.#key || parsedData.key === null)) {
                        this.#responseData.push(parsedData);
                    }
                } catch (jsonError) {
                    // Ignore lines that aren't valid JSON (e.g., debug prints)
                }
            }

            if (this.#responseData.length !== 0) {
                let failure = false;
                for (const resp of this.#responseData) {
                    if (!resp.request_status) {
                        failure = true;
                    }
                }

                if (!failure) {
                    response.request_status = true;
                } else {
                    response.request_status = false;
                }

                response.isUnique = this.#responseData[0].isUnique;

                for (const resp of this.#responseData) {
                    for (const err of resp.errors) {
                        response.errors.push(err);
                    }
                }

                const dataList = [];
                for (const resp of this.#responseData) {
                    dataList.push(resp.data);
                }

                if (response.isUnique) {
                    if (dataList.length === 1) {
                        response.data = dataList[0];
                    } else {
                        response.request_status = false;
                        response.data = null;
                        response.errors.push(`Error: Expected 1 output (isUnique=True) but received ${dataList.length}.`);
                    }
                } else {
                    response.data = dataList;
                }
            } else if (optionalOutput) {
                response.request_status = null;
                response.data = null;
                response.warnings.push("Warning: the output setting is set to optional, and the targeted program didn't gave any output.");
            } else {
                response.request_status = false;
                response.data = null;
                response.errors.push("Error: OutputManager might not be used or not correctly.");
            }

            this.response = response;

        } catch (e) {
            if (e.message.includes('File not found') || 
                e.message.includes('not executable') || 
                e.message.includes('not readable') ||
                e.message.includes('Invalid file')) {
                this.response = {
                    request_status: false,
                    data: null,
                    optionalOutput: optionalOutput,
                    isUnique: isUnique,
                    warnings: ["Warning: targeted file not found or can't be executed, consider checking file informations and language dependencies."],
                    errors: [`Error: ${e.message}`]
                };
            } else {
                this.response = {
                    request_status: false,
                    data: null,
                    optionalOutput: optionalOutput,
                    isUnique: isUnique,
                    warnings: [],
                    errors: [`Unexpected error: ${e.message}`]
                };
            }
        }
    }

    /**
     * Get the full response object.
     *
     * @returns {Object} The complete response with status, data, errors, warnings.
     */
    getResponse() {
        return this.response;
    }

    /**
     * Get the response data if request was successful.
     *
     * @returns {*} The data from the response (any type), or null if request failed.
     *              The return type matches the type sent by the target process.
     */
    getData() {
        if (this.response) {
            if (this.response.request_status) {
                return this.response.data;
            }
        }
        return null;
    }
}

/**
 * OutputManager - Manages receiving requests from other processes and sending responses.
 * 
 * This is a class-based/static manager - all methods are static.
 * Must call init() before using.
 * 
 * @property {*} data - The request data (accessible after init())
 */
class OutputManager {
    static #originalStdout = null;
    static #request = null;
    static #data = null;
    static #requestStatus = null;
    static #optional = null;
    static #uniqueState = null;
    static #initError = null;
    static #errors = [];
    static #warnings = [];

    static data = null;

    /**
     * Initialize OutputManager and read request from stdin.
     * 
     * Must be called before using output() or accessing data.
     * Suppresses stdout to prevent pollution of JSON protocol.
     */
    static init() {
        // Save original stdout.write function so we can restore it later
        this.#originalStdout = process.stdout.write.bind(process.stdout);
        
        // Replace stdout.write with no-op to suppress all console.log/print statements
        process.stdout.write = () => {};
        
        // Read the entire stdin (the JSON request from InputManager)
        this.#request = fs.readFileSync(0, 'utf-8');
        this.#data = JSON.parse(this.#request);
        this.data = this.#data.data;
        this.#optional = this.#data.optionalOutput;
        
        // Reset state for new request
        this.#errors = [];
        this.#warnings = [];
        this.#initError = null;
        this.#requestStatus = null;
        this.#uniqueState = null;
    }

    /**
     * Internal method to write JSON response to stdout.
     * 
     * @private
     * @param {*} args - Data to send in response
     * @param {Object} _data - Request data for key/metadata
     */
    static #write(args, _data) {
        // Restore original stdout.write to actually output the response
        if (this.#originalStdout) {
            process.stdout.write = this.#originalStdout;
        }
        
        // Build and write JSON response
        const response = JSON.stringify({
            key: _data ? _data.key : null,
            request_status: this.#requestStatus,
            data: args,
            optionalOutput: this.#optional,
            isUnique: _data ? _data.isUnique : null,
            errors: this.#errors,
            warnings: this.#warnings
        });
        
        process.stdout.write(response + '\n');
    }

    /**
     * Send response back to the calling process.
     * 
     * @param {*} val - Data to send (any JSON-serializable type)
     * 
     * Note:
     *     Can be called multiple times if isUnique=false in request.
     *     Will error if called multiple times when isUnique=true.
     */
    static output(val) {
        // Check if OutputManager was initialized
        if (!this.#data) {
            if (!this.#initError) {
                if (this.#originalStdout) {
                    process.stdout.write = this.#originalStdout;
                }

                this.#requestStatus = false;
                this.#errors.push("Error: OutputManager isn't initialized.");
                this.#write(null, this.#data);
                this.#initError = true;
            }
        } else {
            // Check if we can output based on isUnique setting
            // uniqueState tracks if we've already output once
            if (!this.#uniqueState || !this.#data.isUnique) {
                this.#requestStatus = true;
                this.#write(val, this.#data);
            } else {
                // Multiple outputs when isUnique=true is an error
                this.#requestStatus = false;
                this.#errors.push(`Error: outputs out of bound (isUnique: ${this.#uniqueState}).`);
                this.#write(val, this.#data);
            }

            // Mark that we've output once
            this.#uniqueState = this.#data.isUnique;
            
            // Re-suppress stdout after writing response
            process.stdout.write = () => {};
        }
    }
}

module.exports = { InputManager, OutputManager };
